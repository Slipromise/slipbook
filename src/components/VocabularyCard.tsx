import React, { useState } from "react";
import Card from "react-bootstrap/Card";
import Stack from "react-bootstrap/Stack";
import { isString, isArray } from "lodash";
import styles from "../styles/components/VocabularyCard.module.scss";
import { AiFillSound } from "react-icons/ai";
import { BsChevronDoubleDown, BsChevronDoubleUp } from "react-icons/bs";
type Sentence = { content: string; translation: string } | string;

type Props = {
  name: string;
  ipa: string | string[];
  origin: string | string[];
  partOfSpeech: string | string[];
  definition: string | string[];
  sentences: Sentence | Sentence[];
  onVoice?: () => void;
};

function VocabularyCard({
  name,
  ipa,
  origin,
  partOfSpeech,
  definition,
  sentences,
  onVoice,
}: Props) {
  const [isCollapsed, setIsCollapsed] = useState(() => false);

  return (
    <Card className={styles.container}>
      <Card.Header>
        <Card.Title>{name}</Card.Title>
        <Card.Subtitle>
          {`${isString(partOfSpeech) ? partOfSpeech : partOfSpeech.join(";")} ${
            isString(ipa) ? ipa : ipa.join("")
          }`}
          {onVoice && <AiFillSound size="2em" onClick={onVoice} />}
        </Card.Subtitle>
      </Card.Header>
      <Card.Body>
        <Stack>
          {typeof origin === "string" ? (
            <span>{origin}</span>
          ) : (
            origin.map((item) => <span key={item}>{item}</span>)
          )}
        </Stack>
        <Stack>
          {typeof definition === "string" ? (
            <span>{definition}</span>
          ) : (
            definition.map((d) => <span key={d}>{d}</span>)
          )}
        </Stack>
      </Card.Body>
      <Card.Footer>
        {!isCollapsed && (
          <Stack as="ul">
            {isArray(sentences) ? (
              sentences.map((item) =>
                isString(item) ? (
                  <li key={item}>{item}</li>
                ) : (
                  <li key={item.content}>
                    {item.content}
                    <br />
                    {item.translation}
                  </li>
                )
              )
            ) : isString(sentences) ? (
              <li key={sentences}>{sentences}</li>
            ) : (
              <li key={sentences.content}>
                {sentences.content}
                <br />
                {sentences.translation}
              </li>
            )}
          </Stack>
        )}
        {isCollapsed ? (
          <BsChevronDoubleDown
            onClick={() => setIsCollapsed((b) => !b)}
            className={styles["collapsed-icon"]}
          />
        ) : (
          <BsChevronDoubleUp
            onClick={() => setIsCollapsed((b) => !b)}
            className={styles["collapsed-icon"]}
          />
        )}
      </Card.Footer>
    </Card>
  );
}

export default VocabularyCard;
