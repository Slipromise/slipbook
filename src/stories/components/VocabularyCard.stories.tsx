import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import VocabularyCard from "../../components/VocabularyCard";

const meta: Meta<typeof VocabularyCard> = {
  component: VocabularyCard,
};

export default meta;

type Story = StoryObj<typeof VocabularyCard>;

export const Normal: Story = {
  args: {
    name: "interaction",
    ipa: "[ˌɪntəˈrækʃən]",
    origin: ["inter/act/ion", "between/act/名詞字尾"],
    partOfSpeech: "名詞",
    definition: "互動",
    sentences: {
      content:
        "There's not enough interaction between management and the workers.",
      translation: "管理階層和勞工之間的互動不足",
    },
  },
};
