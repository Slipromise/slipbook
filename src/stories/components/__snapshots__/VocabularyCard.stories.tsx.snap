// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/VocabularyCard Normal play-test 1`] = `
<div class="_container_1j4fq_1 card">
  <div class="card-header">
    <div class="card-title h5">
      interaction
    </div>
    <div class="card-subtitle h6">
      名詞 [ˌɪntəˈrækʃən]
      <svg stroke="currentColor"
           fill="currentColor"
           stroke-width="0"
           viewbox="0 0 1024 1024"
           height="2em"
           width="2em"
           xmlns="http://www.w3.org/2000/svg"
      >
        <path d="M892.1 737.8l-110.3-63.7a15.9 15.9 0 0 0-21.7 5.9l-19.9 34.5c-4.4 7.6-1.8 17.4 5.8 21.8L856.3 800a15.9 15.9 0 0 0 21.7-5.9l19.9-34.5c4.4-7.6 1.7-17.4-5.8-21.8zM760 344a15.9 15.9 0 0 0 21.7 5.9L892 286.2c7.6-4.4 10.2-14.2 5.8-21.8L878 230a15.9 15.9 0 0 0-21.7-5.9L746 287.8a15.99 15.99 0 0 0-5.8 21.8L760 344zm174 132H806c-8.8 0-16 7.2-16 16v40c0 8.8 7.2 16 16 16h128c8.8 0 16-7.2 16-16v-40c0-8.8-7.2-16-16-16zM625.9 115c-5.9 0-11.9 1.6-17.4 5.3L254 352H90c-8.8 0-16 7.2-16 16v288c0 8.8 7.2 16 16 16h164l354.5 231.7c5.5 3.6 11.6 5.3 17.4 5.3 16.7 0 32.1-13.3 32.1-32.1V147.1c0-18.8-15.4-32.1-32.1-32.1z">
        </path>
      </svg>
    </div>
  </div>
  <div class="card-body">
    <div class="vstack">
      <span>
        inter/act/ion
      </span>
      <span>
        between/act/名詞字尾
      </span>
    </div>
    <div class="vstack">
      <span>
        互動
      </span>
    </div>
  </div>
  <div class="card-footer">
    <ul class="vstack">
      <li>
        There's not enough interaction between management and the workers.
        <br>
        管理階層和勞工之間的互動不足
      </li>
    </ul>
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 16 16"
         class="_collapsed-icon_1j4fq_30"
         data-testid="collapseIcon"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill-rule="evenodd"
            d="M7.646 2.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1-.708.708L8 3.707 2.354 9.354a.5.5 0 1 1-.708-.708z"
      >
      </path>
      <path fill-rule="evenodd"
            d="M7.646 6.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1-.708.708L8 7.707l-5.646 5.647a.5.5 0 0 1-.708-.708z"
      >
      </path>
    </svg>
  </div>
</div>
`;
